File: ios/sandwich/AutomationsSandwich.swift
//
//  AutomationsSandwich.swift
//  QonversionSandwich
//
//  Created by Kamo Spertsyan on 13.04.2022.
//  Copyright © 2022 Qonversion Inc. All rights reserved.
//

import Foundation
import Qonversion

public class AutomationsSandwich : NSObject {
  private var automationsEventListener: AutomationsEventListener?

  private var defaultPresentationConfig: Qonversion.ScreenPresentationConfiguration? = nil
  private var screenPresentationConfigs: [String: Qonversion.ScreenPresentationConfiguration] = [:]
  private var isCustomizationDelegateSet = false
  
  @objc public func subscribe(_ automationsEventListener: AutomationsEventListener) {
    self.automationsEventListener = automationsEventListener
    Qonversion.Automations.shared().setDelegate(self)
  }

#if os(iOS)
  @objc public func setScreenPresentationConfig(_ configData: [String: Any], forScreenId screenId: String? = nil) {
    let config = configData.toScreenPresentationConfig()

    if (!isCustomizationDelegateSet) {
      isCustomizationDelegateSet = true
      Qonversion.Automations.shared().setScreenCustomizationDelegate(self)
    }

    if let screenId = screenId {
      screenPresentationConfigs[screenId] = config
    } else {
      screenPresentationConfigs = [:]
      defaultPresentationConfig = config
    }
  }

  @objc public func setNotificationToken(_ token: String) {
    let tokenData: Data = token.toData()
    Qonversion.Automations.shared().setNotificationsToken(tokenData)
  }
  
  @objc public func getNotificationCustomPayload(_ notificationData: [AnyHashable: Any]) -> [AnyHashable: Any]? {
    return Qonversion.Automations.shared().getNotificationCustomPayload(notificationData)
  }
  
  @objc public func handleNotification(_ notificationData: [AnyHashable: Any]) -> Bool {
    return Qonversion.Automations.shared().handleNotification(notificationData)
  }
  
  @objc public func showScreen(_ screenId: String, completion: @escaping BridgeCompletion) {
    Qonversion.Automations.shared().showScreen(withID: screenId) { success, error in
      if let error = error as NSError? {
        return completion(nil, error.toSandwichError())
      }

      completion(["success": success], nil);
    }
  }
#endif

  @objc public func getAvailableEvents() -> [String] {
    let availableEvents: [AutomationsEvent] = [
      .screenShown,
      .actionStarted,
      .actionFailed,
      .actionFinished,
      .automationsFinished
    ]

    return availableEvents.map { $0.rawValue }
  }
}

extension AutomationsSandwich: Qonversion.ScreenCustomizationDelegate {
  public func presentationConfigurationForScreen(_ screenID: String) -> Qonversion.ScreenPresentationConfiguration {
    return screenPresentationConfigs[screenID] ?? defaultPresentationConfig ?? Qonversion.ScreenPresentationConfiguration.default()
  }
}

extension AutomationsSandwich: Qonversion.AutomationsDelegate {
  public func automationsDidShowScreen(_ screenID: String) {
    let payload: BridgeData = ["screenId": screenID]
    automationsEventListener?.automationDidTrigger(event: AutomationsEvent.screenShown.rawValue, payload: payload.clearEmptyValues())
  }
  
  public func automationsDidStartExecuting(actionResult: Qonversion.ActionResult) {
    let payload: BridgeData = actionResult.toMap()
    automationsEventListener?.automationDidTrigger(event: AutomationsEvent.actionStarted.rawValue, payload: payload.clearEmptyValues())
  }
  
  public func automationsDidFailExecuting(actionResult: Qonversion.ActionResult) {
    let payload: BridgeData = actionResult.toMap()
    automationsEventListener?.automationDidTrigger(event: AutomationsEvent.actionFailed.rawValue, payload: payload.clearEmptyValues())
  }
  
  public func automationsDidFinishExecuting(actionResult: Qonversion.ActionResult) {
    let payload: BridgeData = actionResult.toMap()
    automationsEventListener?.automationDidTrigger(event: AutomationsEvent.actionFinished.rawValue, payload: payload.clearEmptyValues())
  }
  
  public func automationsFinished() {
    automationsEventListener?.automationDidTrigger(event: AutomationsEvent.automationsFinished.rawValue, payload: nil)
  }
}

File: android/sandwich/src/main/kotlin/io/qonversion/sandwich/AutomationsSandwich.kt
package io.qonversion.sandwich

import com.qonversion.android.sdk.automations.Automations
import com.qonversion.android.sdk.automations.AutomationsDelegate
import com.qonversion.android.sdk.automations.ScreenCustomizationDelegate
import com.qonversion.android.sdk.automations.dto.QActionResult
import com.qonversion.android.sdk.automations.dto.QScreenPresentationConfig
import com.qonversion.android.sdk.dto.QonversionError
import com.qonversion.android.sdk.listeners.QonversionShowScreenCallback

class AutomationsSandwich {

    private lateinit var automationsDelegate: AutomationsDelegate;

    private var defaultPresentationConfig: QScreenPresentationConfig? = null
    private val screenPresentationConfigs = mutableMapOf<String, QScreenPresentationConfig>()
    private var isCustomizationDelegateSet = false
    private val screenCustomizationDelegate = object : ScreenCustomizationDelegate {
        override fun getPresentationConfigurationForScreen(screenId: String): QScreenPresentationConfig {
            return screenPresentationConfigs[screenId] ?: defaultPresentationConfig ?: QScreenPresentationConfig()
        }
    }

    // region Initialization

    fun setDelegate(eventListener: AutomationsEventListener) {
        automationsDelegate = createAutomationsDelegate(eventListener)
        Automations.shared.setDelegate(automationsDelegate)
    }

    fun setScreenPresentationConfig(configData: Map<String, Any?>, screenId: String? = null) {
        val config = configData.toScreenPresentationConfig()

        if (!isCustomizationDelegateSet) {
            isCustomizationDelegateSet = true
            Automations.shared.setScreenCustomizationDelegate(screenCustomizationDelegate)
        }

        screenId?.let {
            screenPresentationConfigs[screenId] = config
        } ?: run {
            screenPresentationConfigs.clear()
            defaultPresentationConfig = config
        }
    }

    // endregion

    // region Notifications

    fun getNotificationCustomPayload(notificationData: Map<String, Any?>): Map<String, Any?>? {
        val stringData = notificationData.toStringMap()
        return Automations.shared.getNotificationCustomPayload(stringData)
    }

    @Deprecated("Consider removing this method calls. Qonversion is not working with push notifications anymore")
    fun setNotificationToken(token: String) {
        Automations.shared.setNotificationsToken(token)
    }

    @Deprecated("Consider removing this method calls. Qonversion is not working with push notifications anymore")
    fun handleNotification(notificationData: Map<String, Any?>): Boolean {
        val stringData = notificationData.toStringMap()
        return Automations.shared.handleNotification(stringData)
    }

    // endregion

    // region Other

    fun showScreen(screenId: String, resultListener: ResultListener) {
        Automations.shared.showScreen(screenId, object : QonversionShowScreenCallback {
            override fun onSuccess() {
                resultListener.onSuccess(emptyMap())
            }

            override fun onError(error: QonversionError) {
                resultListener.onError(error.toSandwichError())
            }
        });
    }

    // endregion

    // region Private

    private fun createAutomationsDelegate(eventListener: AutomationsEventListener): AutomationsDelegate {
        return object : AutomationsDelegate {
            override fun automationsDidShowScreen(screenId: String) {
                val payload = mapOf("screenId" to screenId)
                eventListener.onAutomationEvent(AutomationsEventListener.Event.ScreenShown, payload)
            }

            override fun automationsDidStartExecuting(actionResult: QActionResult) {
                eventListener.onAutomationEvent(AutomationsEventListener.Event.ActionStarted, actionResult.toMap())
            }

            override fun automationsDidFailExecuting(actionResult: QActionResult) {
                eventListener.onAutomationEvent(AutomationsEventListener.Event.ActionFailed, actionResult.toMap())
            }

            override fun automationsDidFinishExecuting(actionResult: QActionResult) {
                eventListener.onAutomationEvent(AutomationsEventListener.Event.ActionFinished, actionResult.toMap())
            }

            override fun automationsFinished() {
                eventListener.onAutomationEvent(AutomationsEventListener.Event.AutomationsFinished)
            }
        }
    }

    // endregion
}